extensions/data-workspace/src/services/workspaceService.ts

```diff
@@ -265,20 +266,44 @@ export class WorkspaceService implements IWorkspaceService {
 			if (projectTypes && projectTypes.length > 0) {
 				if (projType) {
 					if (projectTypes.findIndex((proj: string) => proj.toUpperCase() === projType) !== -1) {
-						await this.handleProjectProviderExtension(extension);
+						await this.handleProjectProviderExtension(extension, forceActivate);
 						break;
 					}
 				} else {
-					await this.handleProjectProviderExtension(extension);
+					await this.handleProjectProviderExtension(extension, forceActivate);
 				}
 			}
 		}
 	}
 
-	private async handleProjectProviderExtension(extension: vscode.Extension<any>): Promise<void> {
+	/**
+	 * Ensures project provider extension is activated and registered
+	 * @param extension Extension to activate and register
+	 * @param forceActivate Activates the extension if true. If false, only activates the extension if the workspace contains a file with the extension's project type
+	 */
+	private async handleProjectProviderExtension(extension: vscode.Extension<any>, forceActivate: boolean = false): Promise<void> {
 		try {
 			if (!extension.isActive) {
-				await extension.activate();
+				if (forceActivate) {
+					await extension.activate();
+				} else {
+					const projectTypes = extension.packageJSON.contributes?.projects as string[] | undefined;
+					if (!projectTypes) {
+						return;
+					}
+
+					let projFilesInWorkspace: vscode.Uri[] = [];
+					for (const projType of projectTypes) {
+						(await vscode.workspace.findFiles(`**/*.${projType}`)).forEach(f => projFilesInWorkspace.push(f));
```

Comment:
This doesn't currently handle a project being added to the workspace after initial load - for example if they create it with some other extension or outside of VS Code itself. You'll need to do something like set up a file system watcher to handle that.

Reply:
That should get handled by this: https://github.com/microsoft/azuredatastudio/blob/abc2ba41fc03fcea179976eef40ccca6950ed209/extensions/data-workspace/src/main.ts#L35-L37

The `workspaceTreeDataProvider.refresh()` calls `getProjectsInWorkspace()`, which then eventually calls into `handleProjectProviderExtension()` which will activate the project extension if the workspace now has a file with it's contributed project type.

Reply:
Sort of - but that only handles if a new folder is added. What if the user creates a project in a folder that's already opened?

Reply:
Like if they create the project outside of ADS/vscode? The projects extension would get activated when they first go to the database projects viewlet, but they'd have to use the refresh button for the new project to show up the first time if the extension wasn't already activated. Not sure how common of a scenario this would be though, so it might be unnecessary overhead to have a watcher on every folder in the workspace and check for project files every time there's a change.  

If they create the project in ADS/vscode, then the projects extension will get activated by the new/open/create project from db commands, which is how I expect most people to create or open projects, besides opening a folder that contains a project.

Reply:
Alright, as long as you're ok with the possible limitations there. I agree it seems like more of an edge case.